### PR TITLE
WIP: static libraries handling

### PR DESCRIPTION
**Pull request by Alexandr-Solovev**
## Description

_Add a comprehensive description of proposed changes_

_List associated issue number(s) if exist(s): #6 (for example)_

_Documentation PR (if needed): #1340 (for example)_

_Benchmarks PR (if needed): https://github.com/IntelPython/scikit-learn_bench/pull/155 (for example)_

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have added a respective label(s) to PR if I have a permission for that.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance has changed or why changes are not expected.
- [ ] I have provided justification why quality metrics have changed or why changes are not expected.
- [ ] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by napetrov** [comment]
@Alexandr-Solovev  let's move it to folder named rfcs/20250612-static-libs-handling and name file itself as README.md there.

In this case it also would be rendered properly and you could add images and other content related to RFC

**Comment by napetrov** [comment]
Might be elaborate more on several things. 
1. With current use dynamic is considered to be primarily used and what is used in sklearnex.  So static cases are targeted on C++ users looking for removing runtime dependencies. 
2. Mention both Linux and windows cases and that proposal covers both. 
3. How dependency management is proposed to be handled  - defining dependency only for static library packages i.e. dal-static would require mkl-static as an example
4. What need to be done for this RFC - change in build system, dependency updates for existing packages/package managers, documentation/requirements updates, examples updates

File: rfcs/20250612-libs-structure/README.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+- Reduce maintenance burden
+- Align with Intel’s toolchain direction
+- Prevent user-side issues with incompatible linkage setups
+
+## Proposal
+
+We propose to change the structure of static library handling in oneDAL to address several growing concerns:
+
+1. **Remove symbol duplication and overbundling**:  
+   Eliminate the current `addlib` logic that embeds all symbols from backend libraries (e.g., MKL/OpenBLAS) directly into the oneDAL static archives.
```
Comment (david-cortes-intel) [inline rfcs/20250612-libs-structure/README.md]: @Alexandr-Solovev Is the aim here to also eliminate static linkage of standard (=present in netlib) CPU-based BLAS/LAPACK functions from MKL? That would be very problematic for the PyPI distribution of sklearnex, and would require some changes in the conda distribution to make it work, plus would make this conda distribution a lot less compatible with installed user environments due to interdependencies and bugs that block upgrades.

File: rfcs/20250612-libs-structure/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+     - Better compatibility with Intel DPC++ compiler and runtime deployment
+
+   This means that going forward, **`libonedal_dpc.a` will no longer be available** — instead, only dynamic SYCL (`libonedal_sycl.so`) will be supported.  
+   Any applications depending on static SYCL will need to migrate to dynamic linkage.
+
+
+## Open Questions
+
+- How will this change impact current downstream consumers who rely on bundled behavior?
+- Should we maintain backward compatibility (e.g., via a CMake switch)?
```
Comment (david-cortes-intel) [inline rfcs/20250612-libs-structure/README.md]: Which cmake scripts does this refer to?

File: rfcs/20250612-libs-structure/README.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+|   custom  |  710mb*| 3min 13s -j 112              | 15 s                                      |
+*may be changed in the final version
+
+## Introduction
+
+Currently, in the oneDAL build process, mathematical backends such as **OpenBLAS** for ARM and **Intel MKL** are linked by **copying all symbols** from the corresponding static libraries (e.g., `libopenblas.a`, `libmkl_core.a`) into the resulting archive via custom `addlib` logic in the `Makefile`.
+
+This approach leads to bundling large static libraries into our own library archives, increasing binary size and leading to symbol duplication or conflicts if the final application also links to the same libraries.
+
+In addition, the upcoming oneMKL releases will **remove support for static SYCL libraries** (`libmkl_sycl*.a`). To remain compatible with the future oneMKL toolchain and simplify our own build system, we propose to **drop static SYCL support in oneDAL** as well.
```
Comment (napetrov) [inline rfcs/20250612-libs-structure/README.md]: i think we should separate proposals to make them clear - in fact i don't think there is need for SYCL one as currently it impact no other parties. 

The core gist of this proposal is idea of changing way how we handle static cases for oneDAL - instead of addlib math backends - define build time dependency on them

File: rfcs/20250612-libs-structure/README.md (hunk 2 above)
Comment (Alexandr-Solovev) [reply rfcs/20250612-libs-structure/README.md]: https://github.com/uxlfoundation/oneDAL/blob/656f57df53fc3ebe8a26e9e2b227b1ec01a5e044/cmake/templates/oneDALConfig.cmake.in

File: rfcs/20250612-static-libs-handling/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+* What problem do we solve?
+  * **Significant reduction in library size**:  
+    Removing symbol copying for static libraries will reduce the size of the oneDAL standalone package by approximately:
+    - ~50% (~700 MB) on Linux
+    - ~X% (~X GB) on Windows (TBD)
+  * **Cleaner and more maintainable build system**:  
+    Simplifies both Bazel and CMake logic by removing fragile `addlib`-style manual bundling.
+
+* What is the customer impact if we don't do this?
+  * Users may face issues with dependencies handling.
```
Comment (Vika-F) [inline rfcs/20250612-static-libs-handling/README.md:19]: Examples needed here. Which particular issues would users face?

It would be good to provide the link lines "before" and "after" for some particular configuration, Linux + oneMKL, for example.

File: rfcs/20250612-static-libs-handling/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+1. **Remove symbol duplication and overbundling**:  
+   Eliminate the current `addlib` logic that embeds all symbols from backend libraries (e.g., MKL/OpenBLAS) directly into the oneDAL static archives.  
+   Instead, treat these libraries as **link-time dependencies** only — through `target_link_libraries()` in CMake or `EXTRA_LIBS` in Makefile — shifting responsibility for final linkage to the user.  
+   This avoids binary bloat, reduces symbol conflicts, and simplifies maintenance.
+
+## Open Questions
+
+- How will this change impact current downstream consumers who rely on bundled behavior?
+- Should we maintain backward compatibility (e.g., via a CMake switch)?
+- Should we provide guidance or tooling for users to configure correct BLAS/LAPACK backend linkage?
```
Comment (rakshithgb-fujitsu) [inline rfcs/20250612-static-libs-handling/README.md:49]: I agree on this point, rather than providing backward compatibility it’s better to provide documentation on how to link.
